### PR TITLE
Bugfix/as 1559 missing changed label

### PR DESCRIPTION
**Pull request by lvnhmd**
Add 'change' labels on the protocol and protocol sections.

### CODE REVIEW

File: client/helpers/index.js [hunk 1]
```diff
@@ -75,3 +76,12 @@ export const flattenReveals = (fields, values) => {
     ])
   }, []);
 };
+
+export const changed = (fields, latest, granted) => {
+  if(fields.some(k=> latest.some(l=> l.includes(k)))) {
```
Comment (joefitter) [inline client/helpers/index.js]: space after `if`, space after `k`, space after `l` - lint should be picking this up 🤔

**Comment by lennym** [comment]
What was the cause of the bug and how does your code fix it?

File: client/pages/sections/protocols/sections.js [hunk 2]
```diff
@@ -112,25 +113,46 @@ const sortGranted = (sections, isGranted) => (a, b) => {
...
+        <Fragment key={section}>
+        {
+          changedBadge(changed(sections[section].fields.map(field => field.name), props.latest, props.granted))
+        }
         <ExpandingPanel alwaysUpdate={section === 'conditions' || section === 'authorisations'} key={section} title={getTitle(sections[section], newComments, props.values)}>
           {
             getSection(section, { ...props, protocolState, editable, ...sections[section], sectionsLength: size(sections), sectionIndex })
           }
         </ExpandingPanel>
+        </Fragment>
```
Comment (lennym) [inline client/pages/sections/protocols/sections.js]: The indentation here is out of whack somehow. There shouldn't be two closing tags at the same level.

File: client/helpers/index.js [hunk 3]
```diff
@@ -75,3 +76,12 @@ export const flattenReveals = (fields, values) => {
...
+    return LATEST;
+  }
+  if(fields.some(k=> granted.some(l=> l.includes(k)))) {
+    return GRANTED;
+  }
+}
```
Comment (joefitter) [inline client/helpers/index.js]: Not sure this is the right place for this "helper" - it is only used once and feels like it should be just defined where it is used - I would suggest creating a <Changed /> or <HasChanged /> or something component and using this wherever we show a changed label

File: client/components/changed-badge.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+
+class ChangedBadge extends React.Component {
+  render() {
+    if (this.props.fields.some( k=> this.props.latest.some( l=> l.includes(k)))) {
```
Comment (joefitter) [inline client/components/changed-badge.js]: spacing still off here, this line should be
```js
if (this.props.fields.some(k => this.props.latest.some(l => l.includes(k)))) {
```

File: client/components/changed-badge.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      return <span className="badge changed">changed</span>;
+    }
+    if (this.props.fields.some( k=> this.props.granted.some( l=> l.includes(k)))) {
+      return <span className="badge">amended</span>;
+    }
+    return null;
+  }
+}
+
+export default ChangedBadge;
```
Comment (joefitter) [inline client/components/changed-badge.js]: This file doesn't need a fully fledged React.Component - it doesn't have any lifecycle hooks etc so a plain old function would be better in this case.

File: client/components/changed-badge.js (hunk 4 above)
Comment (joefitter) [reply client/components/changed-badge.js]: also what is `k`? Why not `f`, or even better, `field`

File: client/pages/sections/protocols/sections.js [hunk 6]
```diff
@@ -139,7 +148,9 @@ const mapStateToProps = ({
   // show all sections for legacy
   sections: isGranted && schemaVersion !== 0
     ? pickBy(sections, section => section.granted)
-    : sections
+    : sections,
+  latest,
+  granted
```
Comment (joefitter) [inline client/pages/sections/protocols/sections.js]: this file no longer cares about latest/granted - why don't we connect them in the new component you made?

File: client/pages/sections/protocols/protocol-sections.js [hunk 7]
```diff
@@ -46,7 +47,9 @@ class ProtocolSections extends PureComponent {
       editable,
       newComments,
       readonly,
-      schemaVersion
+      schemaVersion,
+      latest,
+      granted
```
Comment (joefitter) [inline client/pages/sections/protocols/protocol-sections.js]: same here, we should be getting these props in the downstream component - `ChangedBadge`

File: client/components/changed-badge.js [hunk 8]
```diff
@@ -1,15 +1,16 @@
...
-class ChangedBadge extends React.Component {
-  render() {
-    if (this.props.fields.some( k=> this.props.latest.some( l=> l.includes(k)))) {
-      return <span className="badge changed">changed</span>;
-    }
-    if (this.props.fields.some( k=> this.props.granted.some( l=> l.includes(k)))) {
-      return <span className="badge">amended</span>;
-    }
-    return null;
+const ChangedBadge = ({ fields, latest, granted }) => {
```
Comment (joefitter) [inline client/components/changed-badge.js]: I would probably default fields to an empty array here, as it will throw an error if undefined for any reason

**Comment by joefitter** [comment]
@lvnhmd This look much better - one thing I would say is you only seem to be using the new component in protocols - can it be used on the application summary too?

**Comment by lvnhmd** [comment]
> @lvnhmd This look much better - one thing I would say is you only seem to be using the new component in protocols - can it be used on the application summary too?

@joefitter - it most certainly can and it is now

File: client/components/changed-badge.js (hunk 8 above)
Comment (joefitter) [reply client/components/changed-badge.js]: BUMP

File: client/components/changed-badge.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { connect } from 'react-redux';
+
+const ChangedBadge = ({ fields = [], latest = [], granted = [] }) => {
+  if (fields.some(field => latest.some(change => change.includes(field)))) {
```
Comment (joefitter) [inline client/components/changed-badge.js]: i didn't actually look at this code in great detail, but I think you want `_.intersection(fields, changes).length`

Comment (lennym) [reply client/components/changed-badge.js]: I'm totally relaxed about whether or not we leave the "double-some" or use `intersection` (it's broadly the same code under the hood either way), but I have strong feelings about putting this much code _inside_ the condition of an `if`.

It forces the user to understand the code in order to parse the condition, where a descriptively named intermediate value can allow a reader to understand the condition without having to mentally reverse-engineer the code. 

```js
const hasChange = /* whatever implementation you choose */;
if (hasChange) {
  /// stuff
}
```

Comment (lennym) [reply client/components/changed-badge.js]: #bemorephil

File: client/components/application-summary.js [hunk 10]
```diff
@@ -176,15 +149,15 @@ class ApplicationSummary extends React.Component {
                 {
                   subsections.map(key => {
                     const subsection = section.subsections[key];
+                    let fields = Object.values(this.props.fieldsBySection[key] || []);
+                    if(subsection.repeats) fields.push(subsection.repeats);
```
Comment (lennym) [inline client/components/application-summary.js]: Can we always use curly braces for conditionals please?

File: client/components/application-summary.js [hunk 11]
```diff
@@ -176,15 +149,15 @@ class ApplicationSummary extends React.Component {
                 {
                   subsections.map(key => {
                     const subsection = section.subsections[key];
+                    let fields = Object.values(this.props.fieldsBySection[key] || []);
```
Comment (lennym) [inline client/components/application-summary.js]: Should be `const` - adding and removing elements of an array doesn't change the value of the `const`.
```suggestion
                    const fields = Object.values(this.props.fieldsBySection[key] || []);
```

File: client/components/changed-badge.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { connect } from 'react-redux';
+
+const ChangedBadge = ({ fields = [], latest = [], granted = [] }) => {
+  const hasChangedFromLatest = fields.some(field => latest.some(change => change.includes(field)));
```
Comment (lennym) [inline client/components/changed-badge.js]: I'd add a check that there are actually _any_ changes before iterating over the fields here. That way we don't end up doing a bunch of iterations that we _know_ are not required.

```suggestion
  const hasChangedFromLatest = latest.length && fields.some(field => latest.some(change => change.includes(field)));
```

File: client/components/changed-badge.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+  if (hasChangedFromLatest) {
+    return <span className="badge changed">changed</span>;
+  }
+  const hasChangedFromGranted = fields.some(field => granted.some(change => change.includes(field)));
```
Comment (lennym) [inline client/components/changed-badge.js]: As above.
```suggestion
  const hasChangedFromGranted = granted.length && fields.some(field => granted.some(change => change.includes(field)));
```

**Pull request merged**